ProjectVoyage/Cell/MessageCell.m

```diff
@@ -0,0 +1,152 @@
+//
+//  MessageCell.m
+//  Whatsapp
+//
+//  Created by Guilherme David on 07/08/22.
+//  Adapted from Rafael Castro
+//
+
+#import "MessageCell.h"
+#import "Message.h"
+
+@interface MessageCell ()
+
+@property (nonatomic, strong) UITextView *textView;
+@property (nonatomic, strong) UIImageView *bubbleImage;
+@end
+
+#define LAYER_HEIGHT 25
+#define NUM_LAYERS 4
+
+@implementation MessageCell
+
+-(CGFloat) minHeight {
+    return LAYER_HEIGHT * NUM_LAYERS;
+}
+
+#pragma mark - Initialization
+
+- (id) init {
+    self = [super init];
+    if (self) {
+        [self commonInit];
+    }
+    return self;
+}
+
+- (id) initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
+    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
+    if (self) {
+        [self commonInit];
+    }
+    return self;
+}
+
+- (void) commonInit {
+    self.backgroundColor = [UIColor clearColor];
+    self.contentView.backgroundColor = [UIColor clearColor];
+    self.selectionStyle = UITableViewCellSelectionStyleNone;
+    self.accessoryType = UITableViewCellAccessoryNone;
+    
+    _textView = [UITextView new];
+    _bubbleImage = [UIImageView new];
+    
+    // subviews for message bubble
+    [self.contentView addSubview:_bubbleImage];
+    [self.contentView addSubview:_textView];
+}
+
+- (void) prepareForReuse {
+    [super prepareForReuse];
+    
+    _textView.text = @"";
+    _bubbleImage.image = nil;
+}
+
+- (void) setMessage:(Message *)message {
+    _message = message;
+    [self buildCell];
+}
+
+- (void) buildCell {
+    [self setupTextView];
+    [self setupBubbleView];
+    
+    [self setNeedsLayout];
+}
+
+#pragma mark - TextView
+
+- (void) setupTextView {
+    CGFloat max_width = 0.5 * self.contentView.frame.size.width;
+    _textView.frame = CGRectMake(0, 0, max_width, MAXFLOAT);
+    _textView.font = [UIFont fontWithName:@"Helvetica" size:17.0];
+    _textView.backgroundColor = [UIColor clearColor];
+    _textView.userInteractionEnabled = NO;
+    
+    _textView.text = _message.text;
+    [_textView sizeToFit];
+    
+    CGFloat textView_x;
+    CGFloat textView_y;
+    CGFloat textView_w = _textView.frame.size.width;
+    CGFloat textView_h = _textView.frame.size.height;
+    UIViewAutoresizing autoresizing;
+    
+    if (_message.isSenderMyself) {
+        textView_x = self.contentView.frame.size.width - textView_w - 20;
```

Comment:
Is it possible to pull out any of these magic numbers to defined values to make the ui layout logic easier to follow?